### PR TITLE
First attempt at implementing a DNSMADEEASY provider using their API

### PR DESCRIPTION
**Pull request by MatthewSteeples**
Comments welcome

### CODE REVIEW

**Comment by rayholtz** [comment]
This would be great! I was going to start writing a powershell script to interface with DNSMadeEasy to insert the TXT records, but if this provider is built, all the better!  Let me know if you need help testing!

**Comment by MatthewSteeples** [comment]
Used this properly ourselves today, and successfully renewed our certificates automagically

`Complete-ACMEChallenge $DnsIdentifier -ChallengeType dns-01 -Handler dnsme -HandlerParameters @{ApiKey="abc123"; SecretKey="abc123"}`

**Comment by MatthewSteeples** [comment]
Rebased off master, I think this is probably ready to go in now @ebekker

File: ACMESharp/ACMESharp/ACME/Providers/DnsMadeEasyChallengeHandler.cs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+            {
+                var wr = createRequest(records + "/" + recordId);
+                wr.Method = "DELETE";
+
+                using (var response = wr.GetResponse())
+                { }
+            }
+        }
+
+        private string getRecordId(DnsChallenge dnsChallenge, DomainDetails domainDetails, string records)
```
Comment (ebekker) [inline ACMESharp/ACMESharp/ACME/Providers/DnsMadeEasyChallengeHandler.cs]: Correct the casing of the method as per default conventions.

File: ACMESharp/ACMESharp/ACME/Providers/DnsMadeEasyChallengeHandlerProvider.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using ACMESharp.Ext;
+
+namespace ACMESharp.ACME.Providers
+{
+	/// <summary>
```
Comment (ebekker) [inline ACMESharp/ACMESharp/ACME/Providers/DnsMadeEasyChallengeHandlerProvider.cs]: Update the XML Comments and attribute description with specifics for the DNSME provider.

**Comment by ebekker** [comment]
@MatthewSteeples , thank you for the contribution!  I've requested a few changes in the code to improve the code quality.

**Comment by ebekker** [comment]
@MatthewSteeples, are you able to test out this code against the DME SANDBOX environment (e.g. API endpoint `https://api.sandbox.dnsmadeeasy.com/V2.0`)?

**Comment by MatthewSteeples** [comment]
@ebekker Sorry for the delay. I'll get the code updated this week

Regarding sandbox, any changes you make to DME Sandbox will not be returned to public DNS queries so LE will not be able to query it. You can test that the updates are being made, but only with a production DME account will LE be able to action the changes

**Comment by MatthewSteeples** [comment]
Pushed an update to resolve the merge conflicts, update method names to fit in with existing convention and updated XML docs

**Comment by engageant** [comment]
Hi folks, where are we at with this?  I have a need to use DNS Made Easy as a provider and was about to roll my own when I stumbled upon this.

**Comment by ebekker** [comment]
The docs have also been updated with this, so hopefully everything works for those with DNS-M-E, give it a try and let us know.

**Comment by engageant** [comment]
Forgive me for my limited knowledge, but I cannot seem to get this working with staging.  If I specify my HandlerParamaters as @{ApiKey = "xxxx"; SecretKey = "yyyy"; Staging = "true"} I get an InvalidCastException.  When I omit the Staging option, my PS console crashes without displaying an error.  I can't seem to figure out the correct syntax for the Staging boolean - what am I missing?

**Comment by ebekker** [comment]
@MatthewSteeples  is going to have to step in here to help, I have no access to DME.

**Comment by MatthewSteeples** [comment]
Which version of Windows / Powershell are you running?

**Comment by engageant** [comment]
Server 2016, Powershell 5.1.14393.1770

**Comment by MatthewSteeples** [comment]
@engageant I can't get the same error here locally. Do you have the right kind of DNSME account for API access (it's not on the basic tier). If you do, could you do a Fiddler trace and send me (privately if you want) a Fiddler trace of the requests to/from DNSME?

**Comment by engageant** [comment]
Perhaps my understanding of the 'staging' parameter is incorrect - is that
not for testing in DME's sandbox?

**Comment by MatthewSteeples** [comment]
@engageant Correct. It's more for testing the functionality of the code rather than the certificates that you want to create. Any changes made in DNSME's sandbox **will not** be visible to LE's servers, which means that they won't be able to use those results to produce certificates

**Comment by engageant** [comment]
@MatthewSteeples I'm using a sandbox account for my testing here.  The domain I'm requesting a cert for is set up in the sandbox account. Am I formatting the staging parameter correctly?

**Comment by ebekker** [comment]
@MatthewSteeples was saying that changes made in STAGING will **not** be visible to LE, so you **cannot** make a request for a cert from LE if you're using only DME STAGING.

DME STAGING is just so you can test interaction with DME's APIs only.

**Comment by engageant** [comment]
Right - understood.  My test was failing well before any cert was requested.

**Comment by MatthewSteeples** [comment]
I've just noticed in your variable that Staging may need to be a boolean rather than a string. Can you try $True instead of true in quotes?

Doesn't explain which it's crashing if you omit the variable though

**Pull request merged**